File: app-core/src/main/java/ru/micron/service/FavouriteService.java
package ru.micron.service;

import java.util.UUID;
import java.util.stream.Collectors;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.PageRequest;
import org.springframework.stereotype.Service;
import ru.micron.dto.FavouriteDto;
import ru.micron.dto.PageableData;
import ru.micron.dto.PageableResponse;
import ru.micron.dto.omdb.FilmResponseDto;
import ru.micron.mapper.FilmMapper;
import ru.micron.persistence.repository.FilmRepository;
import ru.micron.persistence.repository.UserRepository;

@Service
@RequiredArgsConstructor
public class FavouriteService {

  private final UserRepository userRepository;
  private final FilmRepository filmRepository;
  private final UserService userService;
  private final FilmMapper filmMapper;

  public PageableResponse<FilmResponseDto> getUserFilmsByUserUuid(UUID uuid, int page, int size) {
    var user = userService.findById(uuid);
    var result = filmRepository.findFilmsByUsers(user, PageRequest.of(page - 1, size))
        .stream()
        .map(filmMapper::toDto).collect(Collectors.toList());
    return new PageableResponse<FilmResponseDto>()
        .setData(result)
        .setPageable(new PageableData(filmRepository.countFilmsByUsers(user), page, size, result.size()));
  }

  public Boolean checkIsFavourite(FavouriteDto dto) {
    return filmRepository.existsFilmByUsersAndImdbID(
        userService.findById(dto.getUserId()),
        dto.getImdbId()
    );
  }

  public void addFavourite(FavouriteDto dto) {
    var film = filmRepository.getById(dto.getImdbId());
    var user = userRepository.getById(dto.getUserId())
        .addToFavourite(film);
    userRepository.save(user);
  }

  public void removeFavourite(FavouriteDto dto) {
    var film = filmRepository.getById(dto.getImdbId());
    var user = userRepository.getById(dto.getUserId())
        .removeFromFavourite(film);
    userRepository.save(user);
  }
}

File: app-core/src/main/kotlin/ru/micron/service/FavouriteService.kt
package ru.micron.service

import org.springframework.data.domain.PageRequest
import org.springframework.stereotype.Service
import ru.micron.dto.FavouriteDto
import ru.micron.dto.PageableData
import ru.micron.dto.PageableResponse
import ru.micron.dto.omdb.FilmResponseDto
import ru.micron.mapper.FilmMapper
import ru.micron.persistence.model.Film
import ru.micron.persistence.repository.FilmRepository
import ru.micron.persistence.repository.UserRepository
import java.util.*

@Service
class FavouriteService(
    private val userRepository: UserRepository,
    private val filmRepository: FilmRepository,
    private val userService: UserService,
    private val filmMapper: FilmMapper
) {
    fun getUserFilmsByUserUuid(
        uuid: UUID?, page: Int, size: Int
    ): PageableResponse<FilmResponseDto> {
        val user = userService.findById(uuid)
        val result = filmRepository.findFilmsByUsers(user, PageRequest.of(page - 1, size))
        val list = result.content.map { film: Film? -> filmMapper.toDto(film) }
        return PageableResponse(
            list,
            PageableData(
                result.totalElements,
                page.toLong(),
                size.toLong(),
                result.size.toLong()
            )
        )
    }

    fun checkIsFavourite(dto: FavouriteDto): Boolean {
        return filmRepository.existsFilmByUsersAndImdbID(
            userService.findById(dto.userId),
            dto.imdbId
        )
    }

    fun addFavourite(dto: FavouriteDto) {
        val film = filmRepository.getById(dto.imdbId)
        val user = userRepository.getById(dto.userId)
            .addToFavourite(film)
        userRepository.save(user)
    }

    fun removeFavourite(dto: FavouriteDto) {
        val film = filmRepository.getById(dto.imdbId)
        val user = userRepository.getById(dto.userId)
            .removeFromFavourite(film)
        userRepository.save(user)
    }
}